v2/cmd/asoctl/internal/importing/resource_importer.go

```diff
@@ -160,94 +121,220 @@ func (ri *ResourceImporter) Import(
 		resources = append(resources, importer.Resource())
 	}
 
-	ri.report.WriteToLog(ri.log)
-
 	return &ResourceImportResult{
 		resources: resources,
-	}, kerrors.NewAggregate(errs)
+	}, nil
 }
 
-func (ri *ResourceImporter) ImportResource(ctx context.Context, rsrc ImportableResource) error {
-	// Import it
-	gk := rsrc.GroupKind()
-	name := fmt.Sprintf("%s %s", gk, rsrc.Name())
-	bar := ri.progress.AddBar(
-		0, // zero total because we don't know how much work will be needed
-		mpb.BarRemoveOnComplete(),
-		mpb.PrependDecorators(
-			decor.Name(name, decor.WCSyncSpaceR)),
-		mpb.AppendDecorators(decor.Percentage(decor.WC{W: 5})))
+// queueUniqueImporters reads from the candicates channel, putting each importer onto pending exactly once.
+// This ensures each distinct resource is only imported once, regardless of how many times we
+// encounter it. We also proactively buffer the pending channel to avoid blocking. Any fixed size channel
+// would risk deadlock for a sufficiently large resource graph.
+func (ri *ResourceImporter) queueUniqueImporters(
+	candidates chan ImportableResource,
+	pending chan ImportableResource,
+	progress chan progressDelta,
+) {
+	seen := set.Make[string]()
+	var queue []ImportableResource
```

Comment:
I understand what you're doing here I think: making an infinitely buffered channel.

What I see is:
1. All the candidates, pending, and progress channels are unbuffered.
2. The candidate is written to the `pending` channel from the buffer in `queueUniqueImporters`. As far as I understand, this will block until `importWorker` reads from that channel.
3. `importWorker` reads the `pending` channel and writes to the `completed` channel.
4. `collateResults` reads the `completed` channel and may write back to `candidates`. This is presumably the beginning of a deadlock risk.

Because `collateResults` can block trying to write to `candidates`, but `queueUniqueImporters` may be blocked trying to write to `pending` and `importWorker` will be blocked waiting for `collateResults` to read from `completed`, we're in a deadlock.

As you allude to above, there is no amount of buffering that solves this problem as you don't know how many resources can be discovered in the importResult and must be written into `candidates`.

Is another solution to this problem to move the actual writing of the candidates in `collateResults` into a goroutine?
```
		for _, p := range importResult.pending {
			candidates <- p
		}
```
becomes:
```
		go func() {
			for _, p := range importResult.pending {
				candidates <- p
			}
		}
```

Wondering if this is cleaner than what you have here? Even if not, it might help to explain a bit more about where exactly the deadlock you're avoiding is coming from by highlighting that the problem is that once a resource is completed it tries to enqueue an arbitrarily large number of other resources into `candidates`, which may block because `queueUniqueImporters` is blocking waiting for `importWorker` which is blocking waiting for `collateResults` which is blocked waiting for `queueUniqueImporters`.

Reply:
You've described the deadlock problem well. It boils down to the presence of a cycle in the dependency tree between goroutines.

The suggestion of buffering goroutine was made by @babbageclunk and localizes the funky bit in one place.

Termination of the import is done by tracking whether all pending work has been completed. Making this reliable requires adding all child candidates before marking the parent as complete. I suspect using a goroutine as you suggest would add a race condition causing premature exit, but haven't reasoned through this fully.